From 42c7acd2e017b5014a7df76cbbbdbc22ac2e3346 Mon Sep 17 00:00:00 2001
From: blueh <blueh@marvell.com>
Date: Tue, 30 Oct 2018 13:18:56 +0800
Subject: [PATCH 07/11] nand: nfc: add Toshiba and MXIC NAND support

Sync with u-boot, add support for:
    Toshiba 4GB   - 32Gb - TH58NVG5S0FTAK0
    Toshiba 512MB - 4Gb  - TH58NVG2S3HBAI4
    MXIC    256MB - 2Gb  - MX30LF2G18AC
    MXIC    512MB - 4Gb  - MX30LF4G28AB
---
 drivers/mtd/nand/mvebu_nfc/hal/mvNfc.c | 97 ++++++++++++++++++++++++++++++++++
 1 file changed, 97 insertions(+)

diff --git a/drivers/mtd/nand/mvebu_nfc/hal/mvNfc.c b/drivers/mtd/nand/mvebu_nfc/hal/mvNfc.c
index 53bfb84..69aadb3 100644
--- a/drivers/mtd/nand/mvebu_nfc/hal/mvNfc.c
+++ b/drivers/mtd/nand/mvebu_nfc/hal/mvNfc.c
@@ -274,6 +274,103 @@ typedef enum {
 
 /* Defined Flash Types */
 MV_NFC_FLASH_INFO flashDeviceInfo[] = {
+        {                       /* Toshiba 32Gb - TH58NVG5S0FTAK0*/
+        .tADL = 17,             /* tADL, Address to write data delay, tADL = tALH + tALS*/
+        .tCH = 5,               /* tCH, Enable signal hold time */
+        .tCS = 20,              /* tCS, Enable signal setup time */
+        .tWC = 25,              /* tWC, ND_nWE cycle duration */
+        .tWH = 10,              /* tWH, ND_nWE high duration */
+        .tWP = 12,              /* tWP, ND_nWE pulse time */
+        .tRC = 25,              /* tRC, ND_nRE cycle duration */
+        .tRH = 10,              /* tRH, ND_nRE high duration */
+        .tRP = 12,              /* tRP, ND_nRE pulse width */
+        .tR = 30000,            /* tR = tR+tRR+tWB+1, ND_nWE high to ND_nRE low for read - 25000+20+100+1 */
+        .tWHR = 60,             /* tWHR, ND_nWE high to ND_nRE low delay for status read */
+        .tAR = 10,              /* tAR, ND_ALE low to ND_nRE low delay */
+        .tRHW = 60,             /* tRHW, ND_nRE high to ND_nWE low delay */
+        .pgPrBlk = 64,          /* Pages per block - detected */
+        .pgSz = 4096,           /* Page size */
+        .oobSz = 64,            /* Spare size */
+        .blkNum = 16384,        /* Number of blocks/sectors in the flash */
+        .id = 54680,            /* Device ID 0xDevice,Vendor */
+        .model = "Toshiba 32Gb 4bit",
+        .bb_page = 63,          /* Manufacturer Bad block marking page in block */
+        .flags = NFC_CLOCK_UPSCALE_200M
+        },
+
+        {                       /* Toshiba 4Gb - TH58NVG2S3HBAI4 */
+        .tADL = 17,             /* tADL, Address to write data delay, tADL = tALH (2) + tALS (15) */
+        .tCH = 5,               /* tCH, Enable signal hold time */
+        .tCS = 20,              /* tCS, Enable signal setup time */
+        .tWC = 25,              /* tWC, ND_nWE cycle duration */
+        .tWH = 10,              /* tWH, ND_nWE high duration */
+        .tWP = 12,              /* tWP, ND_nWE pulse time */
+        .tRC = 25,              /* tRC, ND_nRE cycle duration */
+        .tRH = 10,              /* tRH, ND_nRE high duration */
+        .tRP = 12,              /* tRP, ND_nRE pulse width */
+        .tR = 25121,            /* tR = tR+tRR+tWB+1, ND_nWE high to ND_nRE low for read - 25000+20+100+1 */
+        .tWHR = 60,             /* tWHR, ND_nWE high to ND_nRE low delay for status read */
+        .tAR = 10,              /* tAR, ND_ALE low to ND_nRE low delay */
+        .tRHW = 30,             /* tRHW, ND_nRE high to ND_nWE low delay */
+        .pgPrBlk = 64,          /* Pages per block - detected */
+        .pgSz = 2048,           /* Page size */
+        .oobSz = 128,           /* Spare size */
+        .blkNum = 4096,         /* Number of blocks/sectors in the flash */
+        .id = 0xDC98,           /* Device ID 0xDevice,Vendor */
+        .model = "Toshiba 4Gb 8bit",
+        .bb_page = 63,          /* Manufacturer Bad block marking page in block */
+        .flags = NFC_CLOCK_UPSCALE_200M
+        },
+
+        {                       /* MXIC 256MB - 2Gb - MX30LF2G18AC*/
+        .tADL = 15,             /* tADL, Address to write data delay, tADL = tALH + tALS*/
+        .tCH = 5,               /* tCH, Enable signal hold time */
+        .tCS = 15,              /* tCS, Enable signal setup time */
+        .tWC = 20,              /* tWC, ND_nWE cycle duration */
+        .tWH = 7,               /* tWH, ND_nWE high duration */
+        .tWP = 10,              /* tWP, ND_nWE pulse time */
+        .tRC = 20,              /* tRC, ND_nRE cycle duration */
+        .tRH = 7,               /* tRH, ND_nRE high duration */
+        .tRP = 10,              /* tRP, ND_nRE pulse width */
+        .tR = 25000,            /* tR = tR+tRR+tWB+1, ND_nWE high to ND_nRE low for read - 25000+20+100+1 */
+        .tWHR = 60,             /* tWHR, ND_nWE high to ND_nRE low delay for status read */
+        .tAR = 10,              /* tAR, ND_ALE low to ND_nRE low delay */
+        .tRHW = 60,             /* tRHW, ND_nRE high to ND_nWE low delay */
+        .pgPrBlk = 64,          /* Pages per block - detected */
+        .pgSz = 2048,           /* Page size */
+        .oobSz = 64,            /* Spare size */
+        .blkNum = 2048,         /* Number of blocks/sectors in the flash */
+        .id = 0xDAC2,            /* Device ID 0xDevice,Vendor */
+        .model = "MXIC 256MB 4bit",
+        .bb_page = 63,          /* Manufacturer Bad block marking page in block */
+        .flags = NFC_CLOCK_UPSCALE_200M
+        },
+
+        {                       /* MXIC 512MB - 4Gb - MX30LF4G28AB*/
+        .tADL = 15,             /* tADL, Address to write data delay, tADL = tALH + tALS*/
+        .tCH = 5,               /* tCH, Enable signal hold time */
+        .tCS = 15,              /* tCS, Enable signal setup time */
+        .tWC = 20,              /* tWC, ND_nWE cycle duration */
+        .tWH = 7,               /* tWH, ND_nWE high duration */
+        .tWP = 10,              /* tWP, ND_nWE pulse time */
+        .tRC = 20,              /* tRC, ND_nRE cycle duration */
+        .tRH = 7,               /* tRH, ND_nRE high duration */
+        .tRP = 10,              /* tRP, ND_nRE pulse width */
+        .tR = 25000,            /* tR = tR+tRR+tWB+1, ND_nWE high to ND_nRE low for read - 25000+20+100+1 */
+        .tWHR = 60,             /* tWHR, ND_nWE high to ND_nRE low delay for status read */
+        .tAR = 10,              /* tAR, ND_ALE low to ND_nRE low delay */
+        .tRHW = 60,             /* tRHW, ND_nRE high to ND_nWE low delay */
+        .pgPrBlk = 64,          /* Pages per block - detected */
+        .pgSz = 2048,           /* Page size */
+        .oobSz = 112,           /* Spare size */
+        .blkNum = 4096,         /* Number of blocks/sectors in the flash */
+        .id = 0xDCC2,            /* Device ID 0xDevice,Vendor */
+        .model = "MXIC 512MB 8bit",
+        .bb_page = 63,          /* Manufacturer Bad block marking page in block */
+        .flags = NFC_CLOCK_UPSCALE_200M
+        },
+
+
 	{			/* Micron 4Gb */
 	.tADL = 70,		/* tADL, Address to write data delay */
 	.tCH = 5,		/* tCH, Enable signal hold time */
-- 
1.9.1

